src/linker/Linker.Steps/BodySubstituterStep.cs

```diff
@@ -0,0 +1,258 @@
+				return GetSubstitutions (fs);
+			}
+		}
+
+		static XPathDocument GetSubstitutions (Stream substitutions)
+		{
+			using (StreamReader sr = new StreamReader (substitutions)) {
+				return new XPathDocument (new StringReader (sr.ReadToEnd ()));
+			}
+		}
+
+		void ReadSubstitutionFile (XPathDocument document)
+		{
+			XPathNavigator nav = document.CreateNavigator ();
+
+			// Initial structure check
+			if (!nav.MoveToChild ("linker", ""))
+				return;
+
+			// TODO: Add handling for feature
+
+			ProcessAssemblies (nav.SelectChildren ("assembly", ""));
+		}
+
+		void ProcessAssemblies (XPathNodeIterator iterator)
+		{
+			while (iterator.MoveNext ()) {
+				var name = GetAssemblyName (iterator.Current);
+
+				var cache = Context.Resolver.AssemblyCache;
+
+				if (!cache.TryGetValue (name.Name, out AssemblyDefinition assembly)) {
+					Context.LogMessage (MessageImportance.Low, $"Could not match assembly '{name.FullName}' for substitution");
+					continue;
+				}
+
+				ProcessAssembly (assembly, iterator);
+			}
+		}
+
+		void ProcessAssembly (AssemblyDefinition assembly, XPathNodeIterator iterator)
+		{
+			ProcessTypes (assembly, iterator.Current.SelectChildren ("type", ""));
+		}
+
+		void ProcessTypes (AssemblyDefinition assembly, XPathNodeIterator iterator)
+		{
+			while (iterator.MoveNext ()) {
+				XPathNavigator nav = iterator.Current;
+
+				string fullname = GetAttribute (nav, "fullname");
+
+				TypeDefinition type = assembly.MainModule.GetType (fullname);
+
+				if (type == null) {
+					Context.LogMessage (MessageImportance.Low, $"Could not resolve type '{fullname}' for substitution");
+					continue;
+				}
+
+				ProcessType (type, nav);
+			}
+		}
+
+		void ProcessType (TypeDefinition type, XPathNavigator nav)
+		{
+			if (!nav.HasChildren)
+				return;
+
+			XPathNodeIterator methods = nav.SelectChildren ("method", "");
+			if (methods.Count == 0)
+				return;
+
+			ProcessMethods (type, methods);
+		}
+
+		void ProcessMethods (TypeDefinition type, XPathNodeIterator iterator)
+		{
+			while (iterator.MoveNext ())
+				ProcessMethod (type, iterator);
+		}
+
+		void ProcessMethod (TypeDefinition type, XPathNodeIterator iterator)
+		{
+			string signature = GetAttribute (iterator.Current, "signature");
+			if (string.IsNullOrEmpty (signature))
+				return;
+
+			MethodDefinition method = FindMethod (type, signature);
+			if (method == null) {
+				Context.LogMessage (MessageImportance.Low, $"Could not find method '{signature}' for substitution");
+				return;
+			}
+
+			string action = GetAttribute (iterator.Current, "body");
+			switch (action) {
+			case "remove":
+				Annotations.SetAction (method, MethodAction.ConvertToThrow);
+				return;
+			case "stub":
+				string value = GetAttribute (iterator.Current, "value");
+				if (value != "") {
+					if (!TryConvertValue (value, method.ReturnType, out object res)) {
+						Context.LogMessage (MessageImportance.Normal, $"Invalid value for '{signature}' stub");
```

Comment:
I'd argue for throwing instead of logging a message.  If we log a message & ignore the item then the user doesn't get the behavior they are expecting and then they have to waste time figuring out why whatever they intended to have stubbed wasn't stubbed.  Best case scenario they notice the log message and fix it, worst case they waste a lot more time fumbling around.  In either case you are wasting more of the users time than just throwing and telling them what mistake they made.

Reply:
If you are not a fan of that, then my other recommendations would be

* Use `MessageImportance.High` instead.

* Put the LogMessge call into a protected virtual method so that we can choose to change the behavior if we want.

Reply:
Fixed by using MessageImportance.High